src/commonMain/kotlin/fr/acinq/lightning/wire/PaymentOnion.kt

```diff
@@ -386,7 +440,18 @@ object PaymentOnion {
         override fun write(out: Output) = tlvSerializer.write(records, out)
 
         companion object : PerHopPayloadReader<NodeRelayPayload> {
-            override fun read(input: Input): NodeRelayPayload = NodeRelayPayload(tlvSerializer.read(input))
+            override fun read(input: Input): Either<InvalidOnionPayload, NodeRelayPayload> {
+                return PerHopPayload.read(input).flatMap { tlvs ->
+                    when {
+                        tlvs.get<OnionPaymentPayloadTlv.AmountToForward>() == null -> Either.Left(InvalidOnionPayload(OnionPaymentPayloadTlv.AmountToForward.tag, 0))
+                        tlvs.get<OnionPaymentPayloadTlv.OutgoingCltv>() == null -> Either.Left(InvalidOnionPayload(OnionPaymentPayloadTlv.OutgoingCltv.tag, 0))
+                        tlvs.get<OnionPaymentPayloadTlv.OutgoingNodeId>() == null -> Either.Left(InvalidOnionPayload(OnionPaymentPayloadTlv.OutgoingNodeId.tag, 0))
+                        tlvs.get<OnionPaymentPayloadTlv.EncryptedRecipientData>() != null -> Either.Left(InvalidOnionPayload(OnionPaymentPayloadTlv.EncryptedRecipientData.tag, 0))
```

Comment:
Why don't you check the absence of `EncryptedRecipientData` for `FinalPayload.Standard` and `ChannelRelayPayload` too?

Reply:
Good point, I think we currently don't check those in `eclair`, but we could!

Reply:
Added in https://github.com/ACINQ/lightning-kmp/pull/631/commits/99cf4c54514b448987bc10c5ae55f48d4fdbb6ea